palace/utils/dorfler.cpp

```diff
@@ -0,0 +1,168 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "dorfler.hpp"
+
+#include <algorithm>
+#include <numeric>
+#include "utils/communication.hpp"
+
+namespace palace::utils
+{
+
+double ComputeDorflerThreshold(double fraction, const Vector &e)
+{
+  // Precompute the sort and partial sum to make evaluating a candidate partition fast.
+  e.HostRead();  // Pull the data out to the host
+  std::vector<double> estimates(e.begin(), e.end());
+  std::sort(estimates.begin(), estimates.end());
+  MFEM_ASSERT(estimates.size() > 0, "Estimates must be non-empty");
+
+  // Accumulate the squares of the estimates.
+  std::vector<double> sum(estimates.size());
+  for (auto &x : estimates)
+  {
+    x *= x;
+  }
+  std::partial_sum(estimates.begin(), estimates.end(), sum.begin());
+  for (auto &x : estimates)
+  {
+    x = std::sqrt(x);
+  }
+
+  // The pivot is the first point which leaves (1-θ) of the total sum after it.
+  auto pivot = std::lower_bound(sum.begin(), sum.end(), (1 - fraction) * sum.back());
+  auto index = std::distance(sum.begin(), pivot);
+  double error_threshold = estimates[index];
+
+  // Compute the number of elements, and error, marked by threshold value e.
+  auto marked = [&estimates, &sum](double e) -> std::pair<std::size_t, double>
+  {
+    const auto lb = std::lower_bound(estimates.begin(), estimates.end(), e);
+    const auto elems_marked = std::distance(lb, estimates.end());
+    const double error_unmarked =
+        lb != estimates.begin() ? sum[sum.size() - elems_marked - 1] : 0;
+    const double error_marked = sum.back() - error_unmarked;
+    return {elems_marked, error_marked};
+  };
+
+  // Each processor will compute a different threshold: if a given processor has lots of low
+  // error elements, their value will be lower and if a processor has high error, their
+  // value will be higher. Thus using the value from the low error processor will give too
+  // many elements, and using the value from the high error processor will give too few. The
+  // correct threshold value will be an intermediate between the min and max over
+  // processors.
+  auto comm = Mpi::World();
+  const double total_error = [&]()
+  {
+    double total_error = sum.back();
+    Mpi::GlobalSum(1, &total_error, comm);
```

Comment:
Can we combine the reductions which can be throughout this function? A single `MPI_AllReduce` on > 1 variables is much preferred to many reductions on a single variable.